### DISCUSSION TITLE
How to match a more specific rule than regex?

### DISCUSSION THREAD

**Discussion by jetjinser** [tree-sitter/tree-sitter#3794] [Q&A]
## Description

I defined the `text` rule `/[^{}\r\n]+/` and the `math` rule `seq("#", "{", $._node, "}")`.

The `math` rule are more specific than `text`, but `text` will consume `#` when matching. Then `math` fails.

### debug log
```
...
lex_internal state:1, row:0, column:0
  consume character:'h'
  consume character:'e'
  consume character:'l'
  consume character:'l'
  consume character:'o'
  consume character:' '
  consume character:'#'
lexed_lookahead sym:text, size:7
...
```
Retrieved from https://github.com/alt-jinser/tree-sitter-issue/blob/2870555554a1ab9d5a0e499fa2b77e6abefa4037/result.debug.txt#L3-L11, output by `tree-sitter parse show.demo --debug`.

## Link
My [grammar.js](https://github.com/alt-jinser/tree-sitter-issue/blob/regex-eat/grammar.js), demo file: [show.demo](https://github.com/alt-jinser/tree-sitter-issue/blob/regex-eat/show.demo)

---
I don't know what to do, please help me recognize `math` from `text`. Any help would be greatly appreciated! 🙏

**Comment by jetjinser** [comment]
I tried to do this, but this will cause the `text` to be split into several pieces. Is there a better way?

```JavaScript
text: _ => /[^{}#\r\n]+/,
math: $ => choice(seq(
  "#",
  "{", $._node, "}",
), alias("#", $.text)),
```

**Comment by sogaiu** [comment]
May not be that useful of an answer (and possibly even wrong or insufficient (^^;), but IIUC if you want to handle whitespace explicitly, `extras` might need some tweaking and I didn't see it mentioned in the `grammar.js` linked above:

> In addition to the name and rules fields, grammars have a few other optional public fields that influence the behavior of the parser.

>    extras - an array of tokens that may appear anywhere in the language. This is often used for whitespace and comments. The default value of extras is to accept whitespace. To control whitespace explicitly, specify extras: $ => [] in your grammar.

via: https://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl (toward the end of the section)

**Reply by jetjinser** [reply]
Thanks for the tip!
In my use case, I don't quite understand what it has to do with spaces, and the `text` rule correctly matches spaces.

I updated the discussion description to make it more precise what I meant (`text` rule consumed `#`).

**Reply by sogaiu** [reply]
Oh may be I'm confused / mistaken (^^;

Perhaps `[^{}\r\n]` doesn't include whitespace?

**Reply by jetjinser** [reply]
Oops, sorry for ignoring your comment! Discussions are not notified in the GitHub inbox 🙏🙏!

Whether or not I add `extra: $ => []` like you said, the parsing error is there (but the parse results are different) for the same reason.

I don't quite understand where the space roles are in my scenario, I guess the text should include the spaces.
Without the extra field, the `text` still can recognize space. I think `extra` just lets tree-sitter know where to look for extra rules if no matching rules are found.

I don't think managing the spaces manually can solve it either The problem I'm facing is because the text there might be `hello#{in math}`.
I'm not sure if I understand your suggestion, please correct me if I'm wrong!

**Reply by sogaiu** [reply]
I just thought it might be possible that whitespace handling could be one factor contributing to things not working.  The `extras` thing having an influence on whitespace handling has come up before and I was just trying to surface it in case it wasn't a factor being considered.

I'm not sure how the negation (i.e. `[^...]`) works with the whitespace stuff TBH (^^;

Sorry!

---

I did find that with the following tweak:

```diff
diff --git a/grammar.js b/grammar.js
index f2bdd77..71d0347 100644
--- a/grammar.js
+++ b/grammar.js
@@ -11,7 +11,7 @@ module.exports = grammar({
       $.math,
     ),

-    text: _ => /[^{}\r\n]+/,
+    text: _ => /[^#{}\r\n]+/,
     math: $ => seq(
       "#",
       "{", $._node, "}",
```

`tree-sitter parse show.demo` gave a nicer result:

```
(source_file [0, 0] - [1, 0]
  (text [0, 0] - [0, 6])
  (math [0, 6] - [0, 16]
    (text [0, 8] - [0, 15])))
```

May be not quite what you wanted and it doesn't really answer the original question perhaps, but FWIW (^^;
